core/invoker/src/main/scala/whisk/core/containerpool/ContainerPool.scala

```diff
@@ -0,0 +1,219 @@
+
+    val pool = new mutable.HashMap[ActorRef, WorkerData]
+    val prewarmedPool = new mutable.HashMap[ActorRef, WorkerData]
+
+    prewarmConfig.foreach { config =>
+        logging.info(this, s"pre-warming ${config.count} ${config.exec.kind} containers")
+        (1 to config.count).foreach { _ =>
+            prewarmContainer()
+        }
+    }
+
+    def receive: Receive = {
+        // A job to run on a container
+        case r: Run =>
+            // Schedule a job to a warm container
+            ContainerPool.schedule(r.action, r.msg.user.namespace, pool.toMap).orElse {
+                // Create a cold container iff there's space in the pool
+                if (pool.size < poolSize) {
+                    takePrewarmContainer(r.action).orElse {
+                        Some(createContainer())
+                    }
+                } else None
+            }.orElse {
+                // Remove a container and create a new one for the given job
+                ContainerPool.remove(r.msg.user.namespace, pool.toMap).map { toDelete =>
+                    removeContainer(toDelete)
+                    createContainer()
+                }
+            } match {
+                case Some(actor) =>
+                    pool.get(actor) match {
+                        case Some(w) =>
+                            pool.update(actor, WorkerData(w.data, Busy))
+                            actor ! r
+                        case None =>
+                            logging.error(this, "actor data not found")
+                            self ! r
+                    }
+                case None =>
+                    // "reenqueue" the request to find a container at a later point in time
+                    self ! r
+            }
+
+        // Container is free to take more work
+        case NeedWork(data: WarmedData) =>
+            pool.update(sender(), WorkerData(data, Free))
+            feed ! ContainerReleased
+
+        // Container is prewarmed and ready to take work
+        case NeedWork(data: PreWarmedData) =>
+            prewarmedPool.update(sender(), WorkerData(data, Free))
+
+        // Container got removed
+        case ContainerRemoved =>
+            pool.remove(sender())
+    }
+
+    /** Creates a new container and updates state accordingly. */
+    def createContainer() = {
+        val ref = childFactory(context)
+        pool.update(ref, WorkerData(NoData(), Free))
+        ref
+    }
+
+    /** Creates a new prewarmed container */
+    def prewarmContainer() = prewarmConfig.foreach(config => childFactory(context) ! Start(config.exec, config.memoryLimit))
+
+    /**
+     * Takes a prewarm container out of the prewarmed pool
+     * iff a container with a matching kind is found.
+     *
+     * @param kind the kind you want to invoke
+     * @return the container iff found
+     */
+    def takePrewarmContainer(action: ExecutableWhiskAction) = prewarmConfig.flatMap { config =>
```

Comment:
That's right, i need access to the "empty" exec of the config though, as constructing such an Exec is as verbose as just doing a flatMap on the config here anyway.